### PR TITLE
makefile: add gotestsum for verify ci

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Problem Summary: 

Currently, there are 2 problems in TiDB CI environment:

- It is not convenient enough to get the error stack of the failed test cases.
- There is no coverage information for each PR.

### What is changed and how it works?

In this PR, [gotestsum](https://github.com/gotestyourself/gotestsum) is used to generate unit test results in XML format that is friendly to machines. This result can be used for advanced analysis or locating the stack automatically.

The coverage information is also translated to the file in XML format.

The gotest is split into 2 parts, which will be run in parallel to accelerate the test: 
- the test for DDL and executor packages
- the other packages

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by purelind — APPROVED** [review]
lgtm

File: Makefile [hunk 1]
```diff
@@ -132,6 +132,26 @@ gotest: failpoint-enable
 	$(GOTEST) -ldflags '$(TEST_LDFLAGS)' $(EXTRA_TEST_ARGS) -cover $(PACKAGES_TIDB_TESTS) -coverprofile=coverage.txt -check.p true > gotest.log || { $(FAILPOINT_DISABLE); cat 'gotest.log'; exit 1; }
 	@$(FAILPOINT_DISABLE)
 
+gotest_in_verify_ci_part_1: failpoint-enable tools/bin/gotestsum tools/bin/gocov tools/bin/gocov-xml
+	@echo "Running in native mode."
```
Comment (zimulala) [inline Makefile:136]: Do we need to add the information of expensive?

File: Makefile [hunk 2]
```diff
@@ -132,6 +132,26 @@ gotest: failpoint-enable
...
+	@mkdir -p $(TEST_COVERAGE_DIR)
+	@export log_level=info; export TZ='Asia/Shanghai'; \
+	CGO_ENABLED=1 tools/bin/gotestsum --junitfile "$(TEST_COVERAGE_DIR)/tidb-junit-report.xml" -- -v -p $(P) --race \
+	-ldflags '$(TEST_LDFLAGS)' $(EXTRA_TEST_ARGS) -coverprofile="$(TEST_COVERAGE_DIR)/tidb_cov.unit_test.out" \
+	$(PACKAGES_TIDB_TESTS_EXPENSIVE) -check.p true || { $(FAILPOINT_DISABLE); exit 1; }
+	tools/bin/gocov convert "$(TEST_COVERAGE_DIR)/tidb_cov.unit_test.out" | tools/bin/gocov-xml > "$(TEST_COVERAGE_DIR)/tidb-coverage.xml"
+	@$(FAILPOINT_DISABLE)
+
+gotest_in_verify_ci_part_2: failpoint-enable tools/bin/gotestsum tools/bin/gocov tools/bin/gocov-xml
+	@echo "Running in native mode."
```
Comment (zimulala) [inline Makefile:146]: Ditto.

**Review by wjhuang2016 — APPROVED** [review]

**Review by zimulala — COMMENTED** [review]
Can you briefly describe what `gotestsum` does in the description?

File: Makefile [hunk 3]
```diff
@@ -132,6 +132,26 @@ gotest: failpoint-enable
 	$(GOTEST) -ldflags '$(TEST_LDFLAGS)' $(EXTRA_TEST_ARGS) -cover $(PACKAGES_TIDB_TESTS) -coverprofile=coverage.txt -check.p true > gotest.log || { $(FAILPOINT_DISABLE); cat 'gotest.log'; exit 1; }
 	@$(FAILPOINT_DISABLE)
 
+gotest_in_verify_ci_part_1: failpoint-enable tools/bin/gotestsum tools/bin/gocov tools/bin/gocov-xml
+	@echo "Running in native mode."
+	@mkdir -p $(TEST_COVERAGE_DIR)
+	@export log_level=info; export TZ='Asia/Shanghai'; \
+	CGO_ENABLED=1 tools/bin/gotestsum --junitfile "$(TEST_COVERAGE_DIR)/tidb-junit-report.xml" -- -v -p $(P) --race \
+	-ldflags '$(TEST_LDFLAGS)' $(EXTRA_TEST_ARGS) -coverprofile="$(TEST_COVERAGE_DIR)/tidb_cov.unit_test.out" \
+	$(PACKAGES_TIDB_TESTS_EXPENSIVE) -check.p true || { $(FAILPOINT_DISABLE); exit 1; }
```
Comment (zimulala) [inline Makefile:141]: Could we put this logic to Jenkins like the unit test in jenkins：https://github.com/PingCAP-QE/ci/blob/main/jenkins/pipelines/ci/tidb/tidb_ghpr_unit_test.groovy#:~:text=grep%20%22%5C%24%7Bpackage_base%7D/sessionctx%22%20packages.list%20%3E%3E%20%20packages.list.short?

Comment (zhouqiang-cl) [reply Makefile:141]: I think it is hard to maintain in jenkins file and contributors do not know where to modify it.

Comment (purelind) [reply Makefile:141]: I think put his logic in makefile will be more reasonable, it save a lot time when we try to integration it to daily-ci or merge-ci.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**